### PR TITLE
Guarantee oid consistency before metadata touch

### PR DESCRIPTION
**Pull request by gfiorav**
This PR does three things:

+ Guarantees that `UserTable`'s `table_id` and the physical table's `oid` are consistent before checking for physical table existence to avoid false positives (and metadata touch errors).
+ Makes syncs use the proper way to update the `UserTable`'s `table_id` so it propagates correctly (and deleting syncs won't create ghost tables)
+ Adds `fix_oid` code after cartodbfy since schema manipulation changes `table_id` as well.

Testing in staging before calling for review.

### CODE REVIEW

**Comment by gfiorav** [comment]
Hey @juanignaciosl , this is ready for review. Last commit is a hash rework, previous commit tests are :green_apple: 

Thanks!

File: app/models/table.rb [hunk 1]
```diff
@@ -1168,19 +1181,22 @@ def privacy_text_for_vizjson
...
-    @user_table.table_id = self.get_table_id
+    @user_table.table_id = fetch_table_id
   end
 
-  def get_table_id
-    record = owner.in_database.select(:pg_class__oid)
-      .from(:pg_class)
-      .join_table(:inner, :pg_namespace, :oid => :relnamespace)
-      .where(:relkind => 'r', :nspname => owner.database_schema, :relname => name).first
+  def fetch_table_id
```
Comment (juanignaciosl) [inline app/models/table.rb]: How come have you changed the oid extraction logic?

File: app/models/synchronization/adapter.rb [hunk 2]
```diff
@@ -66,39 +66,45 @@ def overwrite(table_name, result)
...
       rescue => exception
-        puts "Sync cartodbfy ERROR: #{exception.message}: #{exception.backtrace.join}"
-        CartoDB.notify_error('Error in sync cartodbfy',
-                             error: exception.backtrace.join('\n'), user_id: user.id, table: table_name)
+        message = exception.message
+        backtrace = exception.backtrace.join('\n')
 
-        update_cdb_tablemetadata(table.name)
+        puts "Sync cartodbfy ERROR: #{message}: #{backtrace}"
+        CartoDB.notify_error('Error in sync cartodbfy',
```
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: Please use the brand new CartoDB::Logger.

File: app/models/synchronization/adapter.rb [hunk 3]
```diff
@@ -66,39 +66,45 @@ def overwrite(table_name, result)
       end
 
       def fix_oid(table_name)
-        actual_oid_from_user_database = database.fetch(%Q{SELECT '#{table_name}'::regclass::oid}).first[:oid]
-        table = ::Table.new(:user_table => ::UserTable.where(name: table_name, user_id: user.id).first)
-        table.table_id = actual_oid_from_user_database.to_i
-        table.save
+        user_table = user.tables.where(name: table_name).first
```
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: `fix_oid` is not related to `Synchronization::Adapter`, it should probably be a `Table` method.

File: app/models/synchronization/adapter.rb [hunk 4]
```diff
@@ -66,39 +66,45 @@ def overwrite(table_name, result)
...
+        backtrace = exception.backtrace.join('\n')
 
-        update_cdb_tablemetadata(table.name)
+        puts "Sync cartodbfy ERROR: #{message}: #{backtrace}"
+        CartoDB.notify_error('Error in sync cartodbfy',
+                             error: backtrace,
+                             user_id: user.id,
+                             table: table_name)
+      ensure
+        fix_oid(table_name)
```
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: Where might we want to fix the oid in case of error? I'm not sure about having it at the ensure, at first sight that might lead to a fix after an unexpected premature error. Not bad either, just proposing a second thought.

File: app/models/synchronization/adapter.rb (hunk 3 above)
Comment (gfiorav) [inline app/models/synchronization/adapter.rb]: Agree. That's why I added `sync_table_id`. `fix_oid` is just a call to that method now. I didn't want to move thing too much around though, given the "urgency" of the situation.

File: app/models/synchronization/adapter.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: In fact the same might happen with the call to `update_cdb_tablemetadata`.

File: app/models/synchronization/adapter.rb (hunk 2 above)
Comment (gfiorav) [inline app/models/synchronization/adapter.rb]: Ah, yeah! You mean the `puts` part right?

File: app/models/synchronization/adapter.rb [hunk 5]
```diff
@@ -66,39 +66,45 @@ def overwrite(table_name, result)
...
+        update_cdb_tablemetadata(table_name)
       end
 
       def update_cdb_tablemetadata(name)
-        qualified_name = "\"#{user.database_schema}\".\"#{name}\""
-        user.in_database(as: :superuser).run(%Q{
-          SELECT CDB_TableMetadataTouch('#{qualified_name}')
-        })
+        user.tables.where(name: name).first.update_cdb_tablemetadata
```
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: There're several lookups of a table by name in this class (three at least), it might indicate there's a need of a method.

File: app/models/table.rb (hunk 1 above)
Comment (gfiorav) [inline app/models/table.rb]: Logic is the same (using new hash syntax), just a name change. Didn't like get where it wasn't a true class attribute getter. Preferred fetch since it's actually fetching the an `oid` from `pg_class`.

File: spec/models/carto/user_table_spec.rb [hunk 6]
```diff
@@ -1,38 +1,35 @@
...
-      user_table.estimated_row_count.should == 999
-      user_table.actual_row_count.should == 1000
-    end
 
+      @user_table.estimated_row_count.should == 999
+      @user_table.actual_row_count.should == 1000
+    end
   end
 
+  it 'should sync table_id with physical table oid' do
```
Comment (juanignaciosl) [inline spec/models/carto/user_table_spec.rb]: Nice :-)

File: app/models/synchronization/adapter.rb (hunk 4 above)
Comment (gfiorav) [inline app/models/synchronization/adapter.rb]: Cartodbfy creates an intermediate table. We *always* want the physical table's oid to be in sync with the UserTable table_id. Same goes for invalidation. If it fails before creating the any intermediate tables the worst case scenarios is you sync in vain since the oid hasn't changed and that you invalidate a table unnecessarily. No big deal.

File: spec/models/table_spec.rb [hunk 7]
```diff
@@ -2416,6 +2416,17 @@ def @data_import.data_source=(filepath)
...
     end
+
+    it 'should query Table estimated an actual row count methods' do
+      ::Table.any_instance.stubs(:estimated_row_count).returns(999)
+      ::Table.any_instance.stubs(:actual_row_count).returns(1000)
+
+      table = new_table(user_id: @user.id)
+      table.save
+
+      table.estimated_row_count.should == 999
```
Comment (juanignaciosl) [inline spec/models/table_spec.rb]: Aren't you checking the stubbed values? :sweat_smile:

File: app/models/synchronization/adapter.rb (hunk 5 above)
Comment (gfiorav) [inline app/models/synchronization/adapter.rb]: Yeah, I could extract that... user model is already fat and I'm not sold on it belonging in this class. What do you think?

File: spec/models/table_spec.rb (hunk 7 above)
Comment (gfiorav) [inline spec/models/table_spec.rb]: I've just moved the test, haven't really thought of it. It does seem useless. :laughing:

File: app/models/table.rb [hunk 8]
```diff
@@ -582,7 +582,20 @@ def remove_table_from_user_database
...
+
+    # TODO: Remove this when table_id is guaranteed to be updated
+    unless real_table_oid == table_id
+      CartoDB.notify_debug('Metadata/Physical table oid mismatch',
+                           trace: caller,
+                           oid: real_table_oid,
+                           table_id: table_id,
+                           name: name)
+
+      self.table_id = real_table_oid
```
Comment (juanignaciosl) [inline app/models/table.rb]: Commenting why is this needed (your PR comment) and even a test about it would be interesting.

File: app/models/synchronization/adapter.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: Yes, using `CartoDB::Logger.error` is simpler.

File: app/models/synchronization/adapter.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: If you think it's safe even on `table.import_to_cartodb`, `table.send :set_the_geom_column!` and `table.import_cleanup` error, fine :+1:

File: app/models/synchronization/adapter.rb [hunk 9]
```diff
@@ -66,39 +66,45 @@ def overwrite(table_name, result)
...
-        table.save
+        user_table = user.tables.where(name: table_name).first
+
+        user_table.sync_table_id
+        user_table.save
       end
 
       def cartodbfy(table_name)
-        table = ::Table.new(:user_table => ::UserTable.where(name: table_name, user_id: user.id).first)
+        table = user.tables.where(name: table_name).first.service
```
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: Same about method location: this has 12 code lines, all of them calling `table.`.

File: app/models/synchronization/adapter.rb (hunk 5 above)
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: I'm not sure about moving it to `User`, but maybe creating a `Adapter` method, meaning "the table related to this synchronization might make sense.

File: app/models/synchronization/adapter.rb (hunk 9 above)
Comment (gfiorav) [inline app/models/synchronization/adapter.rb]: I'd rather not move it around this time. It's delicate. I can create a ticket to refactor this stuff, but let's unblock concurrent ghost tables!

File: app/models/synchronization/adapter.rb (hunk 2 above)
Comment (gfiorav) [inline app/models/synchronization/adapter.rb]: Ok! Wanted to make sure you weren't including the `CartoDB.notify_error` part. Changing!

File: app/models/synchronization/adapter.rb (hunk 4 above)
Comment (gfiorav) [inline app/models/synchronization/adapter.rb]: I think it's safe. It hasn't caused much problems since it was originally written. Calling for a refactor, nevertheless. I'll be glad to do it when CGT is deployed :)

File: app/models/table.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/models/table.rb]: Oh, I was comparing this to `actual_oid_from_user_database = database.fetch(%Q{SELECT '#{table_name}'::regclass::oid}).first[:oid]`. Was this code "incorrect" you fixed?

File: app/models/table.rb (hunk 8 above)
Comment (gfiorav) [inline app/models/table.rb]: Ok, writing a test for that!

**Comment by juanignaciosl** [comment]
I've done some comments, but none of them are major (the only major thing was the "logic change" but I think it was my fault). Although I'd prefer doing some refactors, if you think this is urgent feel free to merge this after testing. Otherwise ping me again after doing the changes.

File: app/models/synchronization/adapter.rb (hunk 5 above)
Comment (gfiorav) [inline app/models/synchronization/adapter.rb]: OK, when you put it like that you convince me :P 

Changing!

File: app/models/synchronization/adapter.rb (hunk 9 above)
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: :+1:

File: app/models/synchronization/adapter.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: Well, I meant replacing both with the Logger call indeed.

File: app/models/synchronization/adapter.rb (hunk 5 above)
Comment (juanignaciosl) [inline app/models/synchronization/adapter.rb]: Ha xD. Not a big deal either. Reading the code I've seen that `Sychronization::Adapter` do too many table-related stuff but again, if you want to deploy this asap feel free to delay this.

File: app/models/table.rb (hunk 1 above)
Comment (gfiorav) [inline app/models/table.rb]: Ah! That's way up there! :stuck_out_tongue:

I think it's better since you don't have to do any casting to get the actual table; just use the `user.tables...`. Also, it uses the `.sync_table_id` which does everything and belongs to `UserTable`.

**Comment by gfiorav** [comment]
Hey @juanignaciosl , thanks for the review. I've only changed to Logger and removed the useless test in `table_spec.rb`. The rest will wait, I've already tested this version and I'm confident the refactors will come in this area (and are not too far away, lot's of duplicate code and error prone cases).

As soon as we're green I'll ping you again.

**Comment by gfiorav** [comment]
@juanignaciosl , green!

Deploying tomorrow if :+1: from your part!

**Comment by juanignaciosl** [comment]
As we've talked about, I'd do some refactors but since this is sensitive delaying it makes sense. I can only say _test it, please_ and :+1: :-)
